### PR TITLE
CacheFilter: handle encodeHeaders while lookups are outstanding

### PR DESCRIPTION
**Pull request by mkbehr**
Commit Message: Handle CacheFilter::encodeHeaders while lookups are outstanding.
Additional Description: CacheFilter stops filter chain iteration while a cache lookup is outstanding, but the filter chain manager may send a local reply anyway, e.g. in case of a stream idle timeout. Handle that case instead of assuming that the lookup has finished.
Risk Level: Low
Testing: Unit test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: test/extensions/filters/http/cache/cache_filter_test.cc [hunk 1]
```diff
@@ -762,6 +762,104 @@ TEST_F(CacheFilterTest, FilterDeletedBeforePostedCallbackExecuted) {
...
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+    filter->onStreamComplete();
+    EXPECT_THAT(lookupStatus(), IsOkAndHolds(LookupStatus::RequestIncomplete));
+    filter->onDestroy();
+    filter.reset();
+  }
+}
+
+using CacheFilterDeathTest = CacheFilterTest;
```
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_test.cc:810]: what does this nickname do?

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 2]
```diff
@@ -102,6 +102,18 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
     return Http::FilterHeadersStatus::Continue;
   }
 
+  if (lookup_result_ == nullptr) {
+    // Filter chain iteration is paused while a lookup is outstanding, but the filter chain manager
+    // can still generate a local reply. One case where this can happen is when a downstream idle
+    // timeout fires, which may mean that the HttpCache isn't correctly setting deadlines on its
+    // asynchronous operations or is otherwise getting stuck.
```
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_filter.cc:109]: are there other cases? It looks like in the timeout case ENVOY_BUG fires, logs, increments a stat or crashes.

In what other case does this block fire?

File: test/extensions/filters/http/cache/cache_filter_test.cc [hunk 3]
```diff
@@ -762,6 +762,104 @@ TEST_F(CacheFilterTest, FilterDeletedBeforePostedCallbackExecuted) {
...
+    // Make sure that the filter doesn't try to encode the cached response after processing the
+    // local reply.
+    EXPECT_CALL(decoder_callbacks_, continueDecoding).Times(0);
+    EXPECT_CALL(decoder_callbacks_, encodeHeaders_).Times(0);
+
+    // Run events on the dispatcher so that the lookup callback is invoked after the local reply.
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+    filter->onStreamComplete();
+    EXPECT_THAT(lookupStatus(), IsOkAndHolds(LookupStatus::RequestIncomplete));
```
Comment (jmarantz) [inline test/extensions/filters/http/cache/cache_filter_test.cc:804]: in this test should we be looking for NotServingFromCache somehow?

**Review by jmarantz — COMMENTED** [review]
lgtm basically; just questions about comments and the test.

File: test/extensions/filters/http/cache/cache_filter_test.cc (hunk 3 above)
Comment (mkbehr) [reply test/extensions/filters/http/cache/cache_filter_test.cc:804]: NotServingFromCache is an implementation detail and not visible to the test, but we exercise its behavior by confirming that the filter doesn't make any calls to the decoder callbacks when the lookup callback comes off the dispatcher.

File: test/extensions/filters/http/cache/cache_filter_test.cc (hunk 1 above)
Comment (mkbehr) [reply test/extensions/filters/http/cache/cache_filter_test.cc:810]: Tells googletest it's a death test for threading purposes. See https://google.github.io/googletest/advanced.html#death-test-naming

File: source/extensions/filters/http/cache/cache_filter.cc (hunk 2 above)
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_filter.cc:109]: This makes sense, if by async you mean same-thread, but after the cache filter has yielded control.

**Review by jmarantz — COMMENTED** [review]
just one comment nit and i'm good to go.

File: test/extensions/filters/http/cache/cache_filter_test.cc (hunk 1 above)
Comment (mkbehr) [reply test/extensions/filters/http/cache/cache_filter_test.cc:812]: Done.

**Review by jmarantz — APPROVED** [review]
thanks!

**Pull request merged**